base_location_geonames_import/wizard/geonames_import.py

```diff
@@ -0,0 +1,269 @@
+                row, self.country_id) if search_states else False
+            if not state:
+                state_vals = self.prepare_state(row, self.country_id)
+                if state_vals not in state_vals_list:
+                    state_vals_list.append(state_vals)
+            else:
+                state_dict[state.code] = state.id
+            if max_import and (i + 1) == max_import:
+                break
+            i += 1
+
+        created_states = self.env['res.country.state'].create(state_vals_list)
+        i = 0
+        for vals in state_vals_list:
+            state_dict[vals['code']] = created_states[i].id
+            i += 1
+        return state_dict
+
+    def _create_cities(self, parsed_csv,
+                       search_cities, max_import, state_dict):
+        # Cities
+        city_vals_list = []
+        city_dict = {}
+        i = 0
+        for row in parsed_csv:
+            city = self.select_city(
+                row, self.country_id) if search_cities else False
+            if not city:
+                state_id = state_dict[
+                    row[self.code_row_index or 4]]
+                city_vals = self.prepare_city(
+                    row, self.country_id, state_id)
+                if city_vals not in city_vals_list:
+                    city_vals_list.append(city_vals)
+            else:
+                city_dict[city.name] = city.id
+            if max_import and (i + 1) == max_import:
+                break
+            i += 1
+
+        created_cities = self.env['res.city'].create(city_vals_list)
+        i = 0
+        for vals in city_vals_list:
+            city_dict[vals['name']] = created_cities[i].id
+            i += 1
+        return city_dict
+
+    @api.multi
+    def run_import(self):
+        self.ensure_one()
+        state_model = self.env['res.country.state']
+        zip_model = self.env['res.city.zip']
+        res_city_model = self.env['res.city']
+
+        # Store current record list
+        current_zips = zip_model.search(
+            [('city_id.country_id', '=', self.country_id.id)])
+        search_zips = True and len(current_zips) > 0 or False
+        current_cities = res_city_model.search(
+            [('country_id', '=', self.country_id.id)])
+        search_cities = True and len(current_cities) > 0 or False
+        current_states = state_model.search(
+            [('country_id', '=', self.country_id.id)])
+        search_states = True and len(current_states) > 0 or False
+
+        parsed_csv = self.get_and_parse_csv()
+        max_import = self.env.context.get('max_import', 0)
+        logger.info('Starting to create the cities and/or city zip entries')
+
+        state_dict = self._create_states(parsed_csv,
+                                         search_states, max_import)
+        city_dict = self._create_cities(parsed_csv,
+                                        search_cities, max_import, state_dict)
+
+        # Zips
+        zip_vals_list = []
```

Comment:
I think this would be better as a 'set()'. Looking at your profiling a massive chunk of time is spent checking list membership.  You can simply remove that check and use `zip_vals_list.add()` and then cast to list at create call. I believe create_multi will accept any iterable anyway, but because you can't be sure on overrides, best to cast as a list.  That will give you 10-15% more speed I expect, maybe more.